### PR TITLE
Funsor based `TraceEnum_ELBO` implementation

### LINKED ISSUE #741: Support enumeration in SVI
Currently, enumeration is supported in MCMC but is not yet tested in SVI. It would be nice to support this feature for SVI too.

### PR DESCRIPTION
**Pull request by ordabayevy**
Adapting from `contrib.funsor.infer.TraceEnum_ELBO`

Fixes #741

### CODE REVIEW

File: test/infer/test_gradient.py [hunk 1]
```diff
@@ -0,0 +1,724 @@
...
+    )
+    expected_loss_fn = lambda params: elbo.loss(
+        random.PRNGKey(0), {}, model, config_enumerate(guide), data, params
+    )
+    expected_loss, expected_grads = jax.value_and_grad(expected_loss_fn)(params)
+
+    # Actual grads averaged over num_particles
+    elbo = infer.Trace_ELBO(
+        num_particles=50000,
+    )
```
Comment (ordabayevy) [inline test/infer/test_gradient.py:131]: I just read the docs for `Trace_ELBO` and it says that it doesn't support non reparametrized samplers. So I tried to use `TraceGraph_ELBO` instead here but got gradients that are shifted by 1 compared to `TraceEnum_ELBO`:

```
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=0.02
E   
E   Mismatched elements: 4 / 4 (100%)
E   Max absolute difference: 1.0056443
E   Max relative difference: 0.47962338
E    x: array([[1.332346, 1.889065],
E          [3.42412 , 1.083667]], dtype=float32)
E    y: array([[2.331621, 2.889788],
E          [4.429764, 2.082466]], dtype=float32)
```

Comment (ordabayevy) [reply test/infer/test_gradient.py:131]: That difference seems to be due to these lines in `TraceGraph_ELBO`:

https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/elbo.py#L687-L688

**Review by fehiepsi — APPROVED** [review]
LGTM per pair discussion sessions. @ordabayevy, please feel free to merge when the changes look good to you. Thank you for adding this feature to numpyro!

File: numpyro/infer/elbo.py [hunk 2]
```diff
@@ -710,3 +716,294 @@ def single_particle_elbo(rng_key):
...
+            model_trace, guide_trace = get_importance_trace_enum(
+                seeded_model,
+                seeded_guide,
+                args,
+                kwargs,
+                param_map,
+                self.max_plate_nesting,
+            )
+            check_model_guide_match(model_trace, guide_trace)
+            _validate_model(model_trace, plate_warning="strict")
```
Comment (ordabayevy) [inline numpyro/infer/elbo.py:872]: Added model and guide validations here.

File: numpyro/infer/elbo.py [hunk 3]
```diff
@@ -710,3 +716,294 @@ def single_particle_elbo(rng_key):
...
+                                        "Expected all enumerated sample sites to share a common scale factor, "
+                                        f"but found different scales at plate('{plate}')."
+                                    )
+                            else:
+                                group_scales[plate] = value
+                    scale = (
+                        reduce(lambda a, b: a * b, group_scales.values())
+                        if group_scales
+                        else None
+                    )
```
Comment (ordabayevy) [inline numpyro/infer/elbo.py:954]: Handling common scaling properly and added validation logic similar to Pyro

File: numpyro/infer/elbo.py [hunk 4]
```diff
@@ -710,3 +716,294 @@ def single_particle_elbo(rng_key):
...
+                    # combine deps
+                    deps = frozenset().union(
+                        *[model_deps[name] for name in group_names]
+                    )
```
Comment (ordabayevy) [inline numpyro/infer/elbo.py:958]: No need to hash factors after refactoring partitioning.

File: numpyro/infer/elbo.py [hunk 5]
```diff
@@ -710,3 +716,294 @@ def single_particle_elbo(rng_key):
...
+                        site = guide_trace[key]
+                        if site["infer"].get("enumerate") == "parallel":
+                            for plate in (
+                                frozenset(site["log_measure"].inputs) & elim_plates
+                            ):
+                                raise ValueError(
+                                    "Expected model enumeration to be no more global than guide enumeration, but found "
+                                    f"model enumeration sites upstream of guide site '{key}' in plate('{plate}')."
+                                    "Try converting some model enumeration sites to guide enumeration sites."
+                                )
```
Comment (ordabayevy) [inline numpyro/infer/elbo.py:970]: Checking model/guide enumeration constraint similar to Pyro. This is more fine-grained than in Pyro and passes tests in `test_enum_elbo.py` that Pyro xfails.

File: test/contrib/test_enum_elbo.py [hunk 6]
```diff
@@ -0,0 +1,2257 @@
...
+    assert_equal(iplate_iplate_grad, plate_iplate_grad, prec=1e-4)
+    assert_equal(iplate_iplate_loss, iplate_plate_loss, prec=1e-4)
+    assert_equal(iplate_iplate_grad, iplate_plate_grad, prec=1e-4)
+    assert_equal(iplate_iplate_loss, plate_plate_loss, prec=1e-4)
+    assert_equal(iplate_iplate_grad, plate_plate_grad, prec=1e-4)
+
+
+@pytest.mark.parametrize("guide_scale", [1])
+@pytest.mark.parametrize("model_scale", [1])
+@pytest.mark.parametrize("outer_vectorized", [False, True])
```
Comment (ordabayevy) [inline test/contrib/test_enum_elbo.py:1596]: `outer_vectorized=True` xfails in Pyro.

File: test/contrib/test_enum_elbo.py [hunk 7]
```diff
@@ -0,0 +1,2257 @@
...
+        return elbo.loss(random.PRNGKey(0), {}, hand_model, guide, params)
+
+    auto_loss, auto_grad = jax.value_and_grad(auto_loss_fn)(params)
+    hand_loss, hand_grad = jax.value_and_grad(hand_loss_fn)(params)
+
+    assert_equal(auto_loss, hand_loss, prec=1e-5)
+    assert_equal(auto_grad, hand_grad, prec=1e-5)
+
+
+def test_elbo_enumerate_plate_5():
```
Comment (ordabayevy) [inline test/contrib/test_enum_elbo.py:574]: This test xfails in Pyro.

**Review by ordabayevy — COMMENTED** [review]
Thanks for reviewing and pair coding on this PR @fehiepsi ! I did some refactoring to handle group scaling properly and added model/guide validation checks. Can you review it again please?

File: numpyro/primitives.py [hunk 8]
```diff
@@ -553,6 +553,11 @@ def process_message(self, msg):
             msg["scale"] = scale * (
                 self.size / self.subsample_size if self.subsample_size else 1
             )
+            plate_to_scale = msg.setdefault("plate_to_scale", {})
+            assert self.name not in plate_to_scale
+            plate_to_scale[self.name] = (
+                self.size / self.subsample_size if self.subsample_size else 1
+            )
```
Comment (fehiepsi) [inline numpyro/primitives.py:560]: Could you move this logic to contrib.funsor.enum_messenger.plate instead?

**Review by fehiepsi — APPROVED** [review]
Awesome work! Thanks Yerdos!

**Comment by ordabayevy** [comment]
Thanks Du!

**Pull request merged**